app/controllers/oaisys/pmh_controller.rb

```diff
@@ -6,21 +6,60 @@ def bad_verb
     end
 
     def identify
+      expect_args(for_verb: :Identify)
     end
 
     def list_sets
+      expect_args(for_verb: :ListSets, exclusive: [:resumptionToken])
     end
 
     def list_metadata_formats
+      expect_args(for_verb: :ListMetadataFormats, optional: [:identifier])
     end
 
     def list_records
+      expect_args(for_verb: :ListRecords, required: [:metadataPrefix], optional: [:from, :until, :set], exclusive: [:resumptionToken])
     end
 
     def get_record
+      expect_args(for_verb: :GetRecord, required: [:identifier, :metadataPrefix])
     end
 
     def list_identifiers
+      expect_args(for_verb: :ListIdentifiers, required: [:metadataPrefix], optional: [:from, :until, :set], exclusive: [:resumptionToken])
+    end
+
+    private
+
+    def expect_args (for_verb:, required: [], optional: [], exclusive: [])
+      arguments = params.except('verb', 'controller', 'action').keys.map(&:to_sym)
+      expected_verb_arguments = required + optional + exclusive
+      unexpected_arguments = (arguments - expected_verb_arguments).present?
+      missing_required_arguments = (required - arguments).present?
+
+      if unexpected_arguments || missing_required_arguments
+        raise BadArgumentException.new(for_verb: for_verb)
+      end
+    end
+
+    rescue_from 'Oaisys::PMHException' do |exception|
+      render template: 'pmh/error.xml.builder', locals: { verb: exception.for_verb, error_code: exception.error_code, error_message: exception.error_message }
+    end
+  end
+
+  class PMHException < StandardError
+    attr_reader :for_verb, :error_code, :error_message
+
+    def initialize(for_verb:, error_code:, error_message:)
+      @for_verb = for_verb
+      @error_code = error_code
+      @error_message = error_message
+    end
+  end
+
+  class BadArgumentException < PMHException
+    def initialize(for_verb:, error_code: :badArgument, error_message: I18n.t('error_messages.illegal_or_missing_arguments'))
```

Comment:
Naming wise, conventionally in Ruby this should be `SuchAndSuchError` or potentially just `SuchAndSuch`. I'm really bad about referring to them colloquially as "exceptions", but technically there's a bit of a difference between the two in Ruby

I don't think it makes sense to treat these as instance variables? It wouldn't really be valid for anyone to be able to instantiate a BadArgumentError with some other error_code or error_message.

 In the vein of "establishing a Gold standard for patterns so that when other people copy the approach we started", I could see this leading to misuse where someone instantiates a BadArgumentError or a PMHError directly with a different code and message for a different type of error, rather than creating a new subclass, which is what we actually want to encourage.

It always pays to design our APIs such that they encourage proper use and disallow doing things we don't want to happen.

Probably the only instance variable or constructor argument BadArgumentError should take is the verb, and error codes and messages can be exposed as immutable methods that don't allow for anyone accidentally reassigning anything. PMHError could be an empty abstract class (it doesn't look like it would make sense for some of the other errors to take a verb argument, to me)

Reply:
Oh and it would probably be good to pull these out into a separate file rather than have them live in the controller.

Reply:
Done.